app/plugin_key_value_store.go

```diff
@@ -97,6 +97,30 @@ func (a *App) GetPluginKey(pluginId string, key string) ([]byte, *model.AppError
 	return nil, nil
 }
 
+func (a *App) ModifyKey(pluginId string, key string, f func(value []byte) ([]byte, error)) *model.AppError {
+	kv, err := a.Srv.Store.Plugin().Get(pluginId, key)
+	if err != nil {
+		if err.StatusCode != http.StatusNotFound {
+			mlog.Error("Failed to query plugin key value", mlog.String("plugin_id", pluginId), mlog.String("key", key), mlog.Err(err))
+			return err
+		}
+	}
+
+	modifiedValue, modifyingErr := f(kv.Value)
+
+	if err != nil {
+		mlog.Error("Failed to modify value with given func", mlog.String("plugin_id", pluginId), mlog.String("key", key), mlog.Err(err))
+		return model.NewAppError("App.AtomicModifyKey", "app.plugin_key_value_store.app_error", nil, modifyingErr.Error(), http.StatusInternalServerError)
+	}
+
+	kv.Value = modifiedValue
+	if _, err := a.Srv.Store.Plugin().SaveOrUpdate(kv); err != nil {
```

Comment:
Hmm, i think it'd be better if we bubbled up the true/false value here to let the caller know if it was a successful set or not.

@hanzei any thoughts on this? Deviates from your original proposal.

Reply:
infact, this should be  `a.Srv.Store.Plugin().SetWithOptions(....)` instead

Reply:
That will still not guarantee full atomicity of this function (between Get and Set calls value could change to a new one and then to the old one again). Extreme example, I know ;)